docs/src/docs/asciidoc/end-to-end-gradle-guide.adoc

```diff
@@ -0,0 +1,871 @@
+    }
+}
+----
+
+[[collect-metadata]]
+== Collecting metadata
+
+When your test/application starts to be a bit more complex things like **reflection**, **resources**, **serialization**, **proxies** or **jni** may be required.
+Since the Native Image has closed world assumption, all of these things must be known in advance during the image build.
+The easiest way how this information can be passed to the Native Image is through metadata config file(s) - depending on the GraalVM version you are using, there could be
+a single `reachability-metadata.json` file (for newer GraalVM versions) or multiple json files (`reflect-config.json`, `resource-config.json`, `proxy-config.json`, `serialization-config.json`, `jni-config.json`).
+To learn more about metadata that Native Image consumes, https://www.graalvm.org/latest/reference-manual/native-image/metadata/[see this].
+
+For example, if you run the test that uses the reflection, and you don't have entry for that reflection call in the metadata config file, your test will fail.
+
+To make your test/application work while using reflection (like in this example) or other metadata, you should either generate metadata configurations or write them manually.
+To generate metadata automatically, you can run your tests (or the main application) with the Native Image Agent, that will collect all the metadata your test/application require.
+To enable the agent (through Native Gradle Plugin) you should either:
+
+- add `-Pagent` flag to the command you are executing
+- or add the following block to `graalvmNative` block in the `build.gradle`:
+
+
+[source,groovy,subs="verbatim,attributes", role="multi-language-sample"]
+----
+agent {
+    enabled = true
+}
+----
+
+[source,kotlin,subs="verbatim,attributes", role="multi-language-sample"]
+----
+agent {
+    enabled.set(true)
+}
+----
+
+[.underline]#To generate the metadata file(s) for your `tests` just run:#
+
+- `./gradlew test` if you added the agent block to the configuration or `./gradlew -Pagent test` if you didn't. This command runs on JVM with native-image-agent and collects the metadata.
+- `./gradlew nativeTest` if you added the agent block to the configuration or `./gradlew -Pagent nativeTest` if you didn't. This command runs on JVM with the native-image agent, collects the metadata and uses it for testing on native-image.
+
+
+[.underline]#To generate the metadata file(s) for your `application` just run:#
+
+- `./gradlew run` if you added the agent block to the configuration or `./gradlew -Pagent run` if you didn't. This command runs on JVM with native-image-agent and collects the metadata.
+
+[WARNING]
+====
+Unless you specify the following block in your `build.gradle` file, Gradle will pick up the Agent from the `JAVA_HOME` environment variable.
+This may cause problems if you set `GRAALVM_HOME` environment variable because your project will be build with one version and the Agent will generate metadata for the other one.
```

Comment:
```suggestion
This may cause problems if you set `GRAALVM_HOME` environment variable because your project will be built with one version and the Agent will generate metadata for the another one.
```